src/zerocoin.cpp

```diff
@@ -91,6 +93,125 @@ std::pair<std::unique_ptr<libzerocoin::CoinSpend>, uint32_t> ParseZerocoinSpend(
+        // we allow transactions of remint type only during specific window
+        return false;
+    
+    // There should only one remint input
+    if (tx.vin.size() != 1 || tx.vin[0].scriptSig.size() == 0 || tx.vin[0].scriptSig[0] != OP_ZEROCOINTOSIGMAREMINT)
+        return false;
+
+    vector<unsigned char> remintSerData(tx.vin[0].scriptSig.begin()+1, tx.vin[0].scriptSig.end());
+    CDataStream inStream1(remintSerData, SER_NETWORK, PROTOCOL_VERSION);
+    sigma::CoinRemintToV3 remint(inStream1);
+
+    if (remint.getMintVersion() != ZEROCOIN_TX_VERSION_2) {
+        LogPrintf("CheckRemintZcoinTransaction: only mint of version 2 is currently supported\n");
+        return false;
+    }
+
+    vector<unique_ptr<sigma::PublicCoin>> sigmaMints;
+    int64_t totalAmountInSigmaMints = 0;
+
+    if (CZerocoinState::IsPublicCoinValueBlacklisted(remint.getPublicCoinValue())) {
+        LogPrintf("CheckRemintZcoinTransaction: coin is blacklisted\n");
+        return false;
+    }
+
+    // All the outputs should be sigma mints
+    for (const CTxOut &out: tx.vout) {
+        if (out.scriptPubKey.size() == 0 || out.scriptPubKey[0] != OP_SIGMAMINT)
+            return false;
+
+        sigma::CoinDenomination d;
+        if (!sigma::IntegerToDenomination(out.nValue, d, state))
+            return false;
+
+        secp_primitives::GroupElement mintPublicValue = sigma::ParseSigmaMintScript(out.scriptPubKey);
+        sigma::PublicCoin *mint = new sigma::PublicCoin(mintPublicValue, d);
+        if (!mint->validate()) {
+            LogPrintf("CheckRemintZcoinTransaction: sigma mint validation failure\n");
+            return false;
+        }
+
+        sigmaMints.emplace_back(mint);
+        totalAmountInSigmaMints += out.nValue;
+    }
+
+    if (remint.getDenomination()*COIN != totalAmountInSigmaMints) {
+        LogPrintf("CheckRemintZcoinTransaction: incorrect amount\n");
+        return false;
+    }
+
+    // Create temporary tx, clear remint signature and get its hash
+    CMutableTransaction tempTx = tx;
+    CDataStream inStream2(remintSerData, SER_NETWORK, PROTOCOL_VERSION);
+    sigma::CoinRemintToV3 tempRemint(inStream2);
+    tempRemint.ClearSignature();
+
+    CDataStream remintWithoutSignature(SER_NETWORK, PROTOCOL_VERSION);
+    remintWithoutSignature << tempRemint;
+
+    CScript remintScriptBeforeSignature;
+    remintScriptBeforeSignature << OP_ZEROCOINTOSIGMAREMINT;
+    remintScriptBeforeSignature.insert(remintScriptBeforeSignature.end(), remintWithoutSignature.begin(), remintWithoutSignature.end());
+
+    tempTx.vin[0].scriptSig = remintScriptBeforeSignature;
+
+    libzerocoin::SpendMetaData metadata(remint.getCoinGroupId(), tempTx.GetHash());
+
+    if (!remint.Verify(metadata)) {
+        LogPrintf("CheckRemintZcoinTransaction: remint input verification failure\n");
+        return false;
+    }
+
+    if (!fStatefulZerocoinCheck)
```

Comment:
If fStatefulZerocoinCheck is set, don't we also need to update the sigmaState, because the outputs are sigma mints?

Reply:
it will be eventually updated because CheckSigmaTransaction is called on the tx too (all the output are in fact regular sigma mints)

Reply:
There is a test that resulting sigma mint is spendable. It wouldn't be otherwise

Reply:
I see, this is because you have made this type of transaction a part of function IsZerocoinV3SigmaTransaction.